### PR TITLE
ISSUE-30954, 30955: fixed issue with cart price rule descriptions

### LINKED ISSUE #30954: Magento 2.4.1 doesn't display Cart Price Rule's description
### Preconditions (*)

1. Pure Magento 2.4.1
2. PHP 7.4.12
3. Mysql - MariaDB 10.4.17
4. Elasticsearch 7.6.2
5. Server - Apache2 (2.4.29)
6. Installed Sample Data

### Steps to reproduce (*)

1. Create cart price rule in admin panel with this set of data and enable it:

> cart rule name: **5% (test)**
> description: **5% test description**
> websites: **Main Website**
> customer groups: **_all available_**
> coupon: **no**
> apply: **percent of product price discount**
> discount amount: **5**
> labels: **_leave empty_**
> 
> **other fields in not necessary.**

2. Go to the storefront catalog and add some product to the cart
3. Go to "View and Edit Cart"

### Expected result (*)

1. Title of discount total contains a description of the applied rule
Screenshot from 2020-11-17 15-08-50

### Actual result (*)

1. Magento doesn't display the description of the created cart rule
Screenshot from 2020-11-17 13-51-17
**_(Magento 2.4.1 displays Cart Price Rule's description if the rule created and applied as COUPON)_**
Screenshot from 2020-11-17 17-57-09

---
Please provide [Severity](https://devdocs.magento.com/guides/v2.3/contributor-guide/contributing.html#backlog) assessment for the Issue as Reporter. This information will help during Confirmation and Issue triage processes.

- [ ] Severity: **S0** _- Affects critical data or functionality and leaves users without workaround._
- [ ] Severity: **S1** _- Affects critical data or functionality and forces users to employ a workaround._
- [x] Severity: **S2** _- Affects non-critical data or functionality and forces users to employ a workaround._
- [ ] Severity: **S3** _- Affects non-critical data or functionality and does not force users to employ a workaround._
- [ ] Severity: **S4** _- Affects aesthetics, professional look and feel, “quality” or “usability”._

### LINKED ISSUE #30955: Magento 2.4.1 displays Cart Price Rule's descriptions incorrectly
### Preconditions (*)

1. Pure Magento 2.4.1
2. PHP 7.4.12
3. Mysql - MariaDB 10.4.17
4. Elasticsearch 7.6.2
5. Server - Apache2 (2.4.29)
6. Installed Sample Data

### Steps to reproduce (*)

1. Create two cart price rules in the admin panel with these sets of data:

**first rule:**
> cart rule name: **5% (test)**
> description: **_leave empty_**
> websites: **Main Website**
> customer groups: **_all available_**
> coupon: **no**
> apply: **percent of product price discount**
> discount amount: **5**
> labels: **_leave empty_**
> 
> **other fields in not necessary.**

**second rule:**
> cart rule name: **10% (test)**
> description: **10% test description**
> websites: **Main Website**
> customer groups: **_all available_**
> coupon: **yes**
> coupone code: **10COUPON**
> apply: **percent of product price discount**
> discount amount: **5**
> labels: **_leave empty_**
> 
> **other fields in not necessary.**

2. Go to the storefront catalog and add **two or more** products to the cart
3. Go to "View and Edit Cart"
4. Apply coupon code "10COUPON"

### Expected result (*)

1. Title of discount total contains only a description of rule with coupon code as rule without code doesn't have a description 
Screenshot from 2020-11-17 16-48-52

### Actual result (*)

1. Title of discount total contains both comma-separated description and code of rule with coupon code
Screenshot from 2020-11-17 16-42-10
2. If you have only one product in the cart and the same rules, then discount descriptions will be displayed correctly
Screenshot from 2020-11-17 17-00-14

---
Please provide [Severity](https://devdocs.magento.com/guides/v2.3/contributor-guide/contributing.html#backlog) assessment for the Issue as Reporter. This information will help during Confirmation and Issue triage processes.

- [ ] Severity: **S0** _- Affects critical data or functionality and leaves users without workaround._
- [ ] Severity: **S1** _- Affects critical data or functionality and forces users to employ a workaround._
- [x] Severity: **S2** _- Affects non-critical data or functionality and forces users to employ a workaround._
- [ ] Severity: **S3** _- Affects non-critical data or functionality and does not force users to employ a workaround._
- [ ] Severity: **S4** _- Affects aesthetics, professional look and feel, “quality” or “usability”._

### PR DESCRIPTION
**Pull request by ivan-prysiazhniuk**
### Description (*)

This PR will fix an issue with Cart Price Rule's description output that described in the issues below

### Fixed Issues (if relevant)

1. Fixes magento/magento2#30954
2. Fixes magento/magento2#30955

### Contribution checklist (*)
 - [x] Pull request has a meaningful description of its purpose
 - [x] All commits are accompanied by meaningful commit messages
 - [x] All new or changed code is covered with unit/integration tests (if applicable)
 - [x] All automated tests passed successfully (all builds are green)

### CODE REVIEW

**Comment by p-ivanko** [comment]
@magento give me test instance

**Comment by p-ivanko** [comment]
@magento run all tests

**Comment by p-ivanko** [comment]
@magento run Functional Tests EE

**Comment by p-ivanko** [comment]
@magento run Integration Tests

**Comment by rogyar** [comment]
Hi @p-ivanko. Thank you for your collaboration, good job. According to the [definition of done](https://devdocs.magento.com/guides/v2.3/contributor-guide/contributing_dod.html) all changes should be covered by auto tests. Could I ask you to cover your changes with an MFTF test, please? The test will go through "steps to reproduce" described in the original issue and assert that the description is not missing. 

Thank you!

**Comment by p-ivanko** [comment]
Hi, @rogyar! I have just created the MTFT test you asked for and pushed it.

File: app/code/Magento/SalesRule/Test/Mftf/Test/StorefrontCartPriceRuleDescriptionTest.xml [hunk 1]
```diff
@@ -0,0 +1,34 @@
...
+        <annotations>
+            <features value="SalesRule"/>
+            <stories value="Create cart price rule"/>
+            <title value="Cart Price Rule description output"/>
+            <description value="Checking the output of the Cart Price Rule description"/>
+            <severity value="MAJOR"/>
+        </annotations>
+
+        <before>
+            <createData entity="ApiCartRule" stepKey="createSalesRule"/>
```
Comment (rogyar) [inline app/code/Magento/SalesRule/Test/Mftf/Test/StorefrontCartPriceRuleDescriptionTest.xml:20]: Hi @p-ivanko. Good job. 
In order to persist test isolation (non-causing side effects by a test for other tests), we need to remove all entities created in the test. The common practice is to do it in the `<after>` section. 

Thanks.

**Review by rogyar — CHANGES_REQUESTED** [review]

File: app/code/Magento/SalesRule/Test/Mftf/Test/StorefrontCartPriceRuleDescriptionTest.xml (hunk 1 above)
Comment (ivan-prysiazhniuk) [reply app/code/Magento/SalesRule/Test/Mftf/Test/StorefrontCartPriceRuleDescriptionTest.xml:20]: @rogyar, oh, sure, sorry.
Done

**Comment by rogyar** [comment]
@magento run all tests

**Comment by p-ivanko** [comment]
@magento run Functional Tests EE Functional Tests B2B

**Comment by p-ivanko** [comment]
@magento run Functional Tests EE

**Comment by p-ivanko** [comment]
@magento run Functional Tests B2B

**Review by rogyar — APPROVED** [review]

**Comment by magento-engcom-team** [comment]
Hi @rogyar, thank you for the review.
[ENGCOM-8563](https://jira.corp.magento.com/browse/ENGCOM-8563) has been created to process this Pull Request

**Comment by engcom-Lima** [comment]
Hi @ivan-prysiazhniuk ,

Thanks for your contribution and collaboration.

Could you please provide the information asked in this [comment](https://github.com/magento/magento2/issues/30955#issuecomment-1333951541).

Thanks

**Comment by engcom-Lima** [comment]
Hi @ivan-prysiazhniuk ,


Thanks for your contribution and collaboration.

Could you please resolve the conflicts in the files, so, that we can proceed further on this PR?

Thanks

**Comment by engcom-Lima** [comment]
Hi @ivan-prysiazhniuk ,

Please do the needful.


Thanks

**Comment by engcom-Dash** [comment]
Closing this PR since it has not been updated for a long time. Please feel free to reopen it if needed.

**Pull request closed without merging**